onnxruntime/core/session/custom_ops.cc

```diff
@@ -119,25 +122,36 @@ common::Status CreateCustomRegistry(const std::vector<OrtCustomOpDomain*>& op_do
     for (const auto* op : domain->custom_ops_) {
       ONNX_NAMESPACE::OpSchema schema(op->GetName(op), "custom op registered at runtime", 0);
 
+      size_t type_id_counter = 0;
       auto input_count = op->GetInputTypeCount(op);
       for (size_t i = 0; i < input_count; i++) {
         auto type = op->GetInputType(op, i);
-        schema.Input(i, "Input" + std::to_string(i), "",
-                     ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED == type
-                         ? "T"
-                         : DataTypeImpl::ToString(onnxruntime::DataTypeImpl::TensorTypeFromONNXEnum(type)));
+        if (ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED == type) {  // Dynamic typed input
+          schema.Input(i, "Input" + std::to_string(i), "", "T" + std::to_string(type_id_counter));
+          schema.TypeConstraint("T" + std::to_string(type_id_counter), DataTypeImpl::ToString(DataTypeImpl::AllTensorTypes()), "all types");
+          type_constraint_ids[op].push_back("T" + std::to_string(type_id_counter++));
+        } else {
+          schema.Input(i, "Input" + std::to_string(i), "", DataTypeImpl::ToString(onnxruntime::DataTypeImpl::TensorTypeFromONNXEnum(type)));
+        }
       }
 
       auto output_count = op->GetOutputTypeCount(op);
       for (size_t i = 0; i < output_count; i++) {
         auto type = op->GetOutputType(op, i);
-        schema.Output(i, "Output" + std::to_string(i), "",
-                      ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED == type
-                          ? "T"
-                          : DataTypeImpl::ToString(onnxruntime::DataTypeImpl::TensorTypeFromONNXEnum(type)));
+        if (ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED == type) {  // Dynamic typed input
+          ORT_ENFORCE(type_id_counter == 1,
+                      "There must be one (and only one) dynamic typed input to the custom op. "
+                      "Its type info at runtime will be used to infer the type info of this dynamic typed output "
+                      "which is required for the success of the model loading step. "
+                      "More than one dynamic typed inputs are currently not supported as differing types at runtime means the output type "
+                      "cannot be inferred without which model loading cannot proceed.");
+
```

Comment:
Why do we do this check here instead of when processing the inputs?

Reply:
Because we want to allow multiple dynamic typed inputs  as long as the  types of the outputs are fixed. We defer checking the count of the dynamic typed inputs until we encounter one dynamic typed output.